src/bin/pg_autoctl/watch.c

```diff
@@ -0,0 +1,1102 @@
+			}
+			else if (context->selectedRow < (context->rows - 6))
+			{
+				context->selectedRow += 5;
+			}
+		}
+
+		/* cancel current selected row */
+		else if (ch == KEY_DL || ch == KEY_DC)
+		{
+			context->selectedRow = 0;
+		}
+	} while (ch != ERR);
+
+	return true;
+}
+
+
+/*
+ * watch_render displays the context on the terminal window.
+ */
+bool
+watch_render(WatchContext *context)
+{
+	int printedRows = 0;
+
+	printedRows += print_watch_header(context, 0);
+
+	(void) clear_line_at(1);
+	++printedRows;
+
+	/* skip empty lines and headers */
+	if (context->selectedRow > 0 && context->selectedRow < 3)
+	{
+		context->selectedRow = 3;
+	}
+
+	int nodeRows = print_nodes_array(context, 2, 0);
+
+	printedRows += nodeRows;
+
+	(void) clear_line_at(2 + nodeRows);
+	++printedRows;
+
+	/* skip empty lines and headers */
+	if (context->selectedRow >= (2 + nodeRows) &&
+		context->selectedRow <= 2 + nodeRows + 1)
+	{
+		context->selectedRow = 2 + nodeRows + 2;
+	}
+
+	printedRows += print_events_array(context, 2 + nodeRows + 1, 0);
+
+	/* clean the remaining rows that we didn't use for displaying events */
+	if (printedRows < context->rows)
+	{
+		for (int r = printedRows; r < context->rows; r++)
+		{
+			(void) clear_line_at(r);
+		}
+	}
+
+	/* now display the footer */
+	(void) print_watch_footer(context);
+
+	refresh();
+
+	return true;
+}
+
+
+/*
+ * print_watch_header prints the first line of the screen, with the current
+ * formation that's being displayed, the number_sync_standbys, and the current
+ * time.
+ */
+static int
+print_watch_header(WatchContext *context, int r)
+{
+	uint64_t now = time(NULL);
+	char timestring[MAXCTIMESIZE] = { 0 };
+
+	/* make sure we start with an empty line */
+	(void) clear_line_at(0);
+
+	/* format the current time to be user-friendly */
+	epoch_to_string(now, timestring);
+
+	/* "Wed Jun 30 21:49:08 1993" -> "21:49:08" */
+	timestring[11 + 8] = '\0';
+
+	clear_line_at(r);
+
+	mvprintw(r, context->cols - 9, "%s", timestring + 11);
+
+	int c = 0;
+
+	mvprintw(r, c, "Formation: "); /* that's 11 chars */
+	c += 11;
+
+	attron(A_BOLD);
+	mvprintw(r, c, "%s", context->formation);
+	attroff(A_BOLD);
+
+	c += strlen(context->formation);
+
+	/*
+	 * Check if we have enough room for a full label here:
+	 *  - add  9 cols for the date at the end of the line
+	 *  - add 18 cols for the label " - Sync Standbys: "
+	 *  - add  3 cols for the number itself (e.g. "1")
+	 */
+	if (context->cols > (c + 9 + 18 + 3))
+	{
+		mvprintw(r, c, " - Sync Standbys: "); /* that's 18 chars */
+		c += 18;
+	}
+	else
+	{
+		mvprintw(r, c, " - nss: "); /* that's 8 chars */
+		c += 8;
+	}
+
+	attron(A_BOLD);
+	mvprintw(r, c, "%d", context->number_sync_standbys);
+	attroff(A_BOLD);
+
+	/* we only use one row */
+	return 1;
+}
+
+
+/*
+ * print_watch_footer prints the last line of the screen, an help message.
+ */
+static int
+print_watch_footer(WatchContext *context)
+{
+	int r = context->rows - 1;
+	char *help = "Press F1 to exit";
```

Comment:
Apparently F1 is common in Ncurses land... or at least in the docs from previous century. ESC is a can of worms, because of the way a qwerty keyboard is wired, ESC then a is the same thing as M-a (or Alt+a depending on how you want to write it). So at the moment we stay away from ESC and Meta / Alt.